sensor_test.py

```diff
@@ -0,0 +1,172 @@
+
+
+def test_set_sensor_reading_interval(
+        get_sensor_info, set_sensor_reading_interval, get_sensor_reading
+):
+    """
+    1. Set sensor reading interval to 1.
+    2. Get sensor info.
+    3. Validate that sensor reading interval is set to interval from Step 1.
+    4. Get sensor reading.
+    5. Wait for interval specified in Step 1.
+    6. Get sensor reading.
+    7. Validate that reading from Step 4 doesn't equal reading from Step 6.
+    """
+    print("Step 1", "Set sensor reading interval to 1")
+    new_sensor_reading_interval = 1
+    set_sensor_reading_interval(new_sensor_reading_interval)
+
+    print("Step 2", "Get sensor info")
+    sensor_info_step2 = get_sensor_info()
+
+    print("Step 3", "Validate that sensor reading interval is set to interval from Step 1")
+    assert sensor_info_step2.get("reading_interval") == new_sensor_reading_interval
+
+    print("Step 4", "Get sensor reading")
+    sensor_reading_step4 = get_sensor_reading()
+
+    print("Step 5", "Wait for interval specified in Step 1")
+    wait(func=get_sensor_reading,
+         condition=lambda x: isinstance(x, dict),
+         tries=5,
+         timeout=new_sensor_reading_interval)
+
+    print("Step 6", "Get sensor reading")
+    sensor_reading_step6 = get_sensor_reading()
+
+    print("Step 7", "Validate that reading from Step 4 doesn't equal reading from Step 6")
+    assert sensor_reading_step6 != sensor_reading_step4
+
+
+# Максимальна версія прошивки сенсора -- 15
+def test_update_sensor_firmware(get_sensor_info, update_sensor_firmware):
+    """
+    1. Get original sensor firmware version.
+    2. Request firmware update.
+    3. Get current sensor firmware version.
+    4. Validate that current firmware version is +1 to original firmware version.
+    5. Repeat steps 1-4 until sensor is at max_firmware_version - 1.
+    6. Update sensor to max firmware version.
+    7. Validate that sensor is at max firmware version.
+    8. Request another firmware update.
+    9. Validate that sensor doesn't update and responds appropriately.
+    10. Validate that sensor firmware version doesn't change if it's at maximum value.
+    """
+    max_firmware_version = 15
+
+    print("Step 1" "Get original sensor firmware version")
+    original_sensor_fw_ver = get_sensor_info().get("firmware_version")
+
+    print("Step 2" "Request firmware update")
+    sensor_fw_upgrade_request = update_sensor_firmware()
+    assert sensor_fw_upgrade_request == "updating"
+    wait(func=get_sensor_info, condition=lambda x: isinstance(x, dict), tries=15, timeout=3)
+
+    print("Step 3" "Get current sensor firmware version")
+    current_sensor_fw_ver = get_sensor_info().get("firmware_version")
+
+    print("Step 4" "Validate that current firmware version is +1 to original firmware version")
+    assert current_sensor_fw_ver == (original_sensor_fw_ver + 1)
+
+    print("Step 5" "Repeat steps 1-4 until sensor is at max_firmware_version - 1")
```

Comment:
Test description is for humans, while the test is for computers. You can optimize this code by starting with the loop right away until you get to `max_firmware_version - 1`. It still does what the test steps say but in a form more suitable for a programming language.

Reply:
Steps 2-5 were united in one. But to satisfy step4 requirement additional check was added